### PR TITLE
Branch Predictor Fixes & Improvements

### PR DESCRIPTION
**Pull request by jiristefan**
Addresses some points from #143

- fence.i instrukction will now cause predictor to flush BTB rows and BHT states
- Even instructions which are Not Taken are added into BTB and updated into BHT
- Unconditional jumps now have a flag in BTB and do not trigger BHT
- Added 3rd predictor widget with additional information not directly related to BHT and BTB
- Added new statistic counting total encountered branch instructions and pipeline flushes
- Restructured predictor and GUI code

### CODE REVIEW

**Comment by ppisa** [comment]
I have tested updated version with my simple test cases and all looks correct. I would wait for @jdupak for code style review and I hope we merge changes soon.

File: src/gui/windows/predictor/predictor_btb_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_btb_dock.cpp:109]: nit: Duplicated code. The initialization is same in both branches. Should be hoisted before the if and merged with the initializer.

Comment (jdupak) [reply src/gui/windows/predictor/predictor_btb_dock.cpp:109]: Never mind. I misunderstood the code.

File: src/gui/windows/predictor/predictor_btb_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_btb_dock.cpp:118]: I would really prefer not to reuse to variable. It actually confused me for a while. Actually, you dont need a variable at all. You could just call `setData` on the result of the `get_...` method.

File: src/gui/windows/predictor/predictor_btb_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_btb_dock.cpp:145]: Same as above. Btw, given this use, it might be good to make a function that takes 4 strings and sets the all of them. Not blocking.

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_info_dock.cpp:21]: Joke: ...and people are suprised why web apps are so popular :D

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_info_dock.cpp:148]: Maybe I am blind, but the two functions look exactly the same.

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_info_dock.cpp:159]: Here

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_info_dock.cpp:177]: Another duplication

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_info_dock.cpp:226]: I would consider making this `&& is_predictor_dynamic` so that we set the value at just one place. Also those two branches could be merged and just set the value to be `is_predictor_dynamic`, but I am not 100% convinced here.

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jdupak) [inline src/gui/windows/predictor/predictor_info_dock.cpp:251]: Duplication

Comment (jdupak) [reply src/gui/windows/predictor/predictor_info_dock.cpp:251]: With the next fn

File: src/machine/predictor.h
Comment (jdupak) [inline src/machine/predictor.h:226]: If we have this method, why is GUI using manual check for dynamic using enum?

**Review by jdupak — APPROVED** [review]
Lookgs mostly good to me. There are some duplications and redundancies but no blockers. Good job, Jirka!

I am fine merging this now and fixing the issues in a subsequest PR. I leave it up to you.

**Comment by ppisa** [comment]
OK, I agree with incremental progress merging. The core changes are minimal and value of predictor is much bigger/usable for teaching in proposed version.  I would like to add predictor disable action to all processor presets is some followup commit.

**Comment by ppisa** [comment]
I have made merge in parallel to your commit
415f5d13c817503831a2153a757266180b11c8b3 Fixed code duplication and table cell handling clarity
I have applied that commit above the master now.

**Comment by jiristefan** [comment]
Thank you, the last commit should hopefully address all the things mentioned in the review. I pushed it without noticing the merge.

File: src/gui/dialogs/new/newdialog.cpp
Comment (jiristefan) [reply src/gui/dialogs/new/newdialog.cpp:404]: Fixed

File: src/gui/windows/predictor/predictor_bht_dock.cpp
Comment (jiristefan) [reply src/gui/windows/predictor/predictor_bht_dock.cpp:94]: Fixed

File: src/gui/windows/predictor/predictor_btb_dock.cpp
Comment (jiristefan) [reply src/gui/windows/predictor/predictor_btb_dock.cpp:118]: Fixed

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jiristefan) [reply src/gui/windows/predictor/predictor_info_dock.cpp:226]: Fixed

File: src/machine/predictor.h
Comment (jiristefan) [reply src/machine/predictor.h:226]: Moved this to separate function

File: src/gui/windows/predictor/predictor_info_dock.cpp
Comment (jiristefan) [reply src/gui/windows/predictor/predictor_info_dock.cpp:251]: Similar with the previous prediction / update widget duplication, the logic is the same, it just works with different widgets eg. "value_event_predict_instruction" VS "value_event_update_instruction".

The ideal solution would be probably to have the widget as a separate class and have it instantiated two times, but I feel like it would not save that much code, and the need for two separate functions to handle the two widgets would still be there.

If you'd like I can move the code into a new class, but I would prefer not to, unless there is a need to re-use it somewhere else also.

Comment (jdupak) [reply src/gui/windows/predictor/predictor_info_dock.cpp:251]: OH, never mind. I am just blind (and tired). No need to over-engineer this. Keeps it as is.

**Pull request merged**
